### ISSUE TITLE
Annotation of `this` in function declarations

### ISSUE THREAD

**Issue by jussi-kalliokoski** [facebook/flow#452]
Currently it seems that you can only annotate the `this` magic reference by putting the function as a method in a class declaration.

However, in the light of the [new bind syntax proposal](https://github.com/zenparsing/es-function-bind), it would make sense to be able to annotate functions that take in `this` as data, e.g.

``` javascript
function head (count) {
    return this.slice(0, count);
}

[1,2,3]::head(2) // [1,2]
```

could be annotated something like:

``` javascript
function head <T> Array<T> -> (count : number) : Array<T> {
  ...
}
```

**Comment by samwgoldman** [comment]
Flow is (mostly) smart about this already without annotations. I'm not clear what this new bind syntax is (do you have a link to a spec?), but I imagine flow will support it the same way that flow currently supports `bind`, `call`, and `apply`.

```javascript
/* @flow */

function head(count) {
  return this.slice(0, count);
}

head.bind([]) // OK
head.bind(1) // bad
```

**Comment by jussi-kalliokoski** [comment]
> I'm not clear what this new bind syntax is (do you have a link to a spec?)

Ah, sorry, forgot the link, fixed.

> Flow is (mostly) smart about this already without annotations

Yes, but it would be nice to be able to annotate this instead, for documentational purposes (easier to understand at a glance and also allows for better error messages, as well as documentation generation), but also for standalone definitions.

**Comment by jussi-kalliokoski** [comment]
Having thought about this a bit more, a simpler syntax would actually be to allow specifying `this` as a parameter, as so:

```javascript
function add (this : number, b : number) : number {
  return this + x;
}

function add (/*: this : number */, b /*: number */) /*: number */ {
  return this + x;
}

/*: (this : number, b: number) => number */
function add (b) {
  return this + b;
}
```

**Comment by avikchaudhuri** [comment]
Agree that we should support `this` to be annotated as a parameter. This requires some parser work too, btw, since `this` is not a valid identifier everywhere.

We need this whenever methods are exported, by the way. It's true Flow is good in inferring `this` types, but it sucks that this information is lost across modules.

Related: constructor functions can be annotated as of https://github.com/facebook/flow/commit/618186ad5ceb66cf7bf009b2591b74ed885d69bd

**Comment by Artazor** [comment]
Hi guys! what about the following examples: https://github.com/Microsoft/TypeScript/issues/1985#issuecomment-127293978
Are they too complex/ugly?

**Comment by izaakschroeder** [comment]
+1

**Comment by danvk** [comment]
Pulling in some material from #968: an example, and a link to how Google Closure does this.

----

There's an error (marked with `<---`) in this snippet which Flow could potentially catch:

```javascript
class RemoteFile {
  url: string;
...
  promiseXHR(xhr: XMLHttpRequest): Q.Promise<[any, Event]> {
    var url = this.url;
    var deferred = Q.defer();
    xhr.addEventListener('load', function(e) {
      if (this.status >= 400) {
        deferred.reject(this.status + ' ' + this.statusText);
      } else {
        deferred.resolve([this.response, e]);
      }
    });
    xhr.addEventListener('error', function(e) {
      deferred.reject(`Request for ${this.url} failed: ${this.status}`);  // <---
    });
    this.numNetworkRequests++;
    xhr.send();
    return deferred.promise;
  }
}
```

The problem is that `this` inside the `addEventListener` callback refers to the `XMLHttpRequest` object (I believe) and not the `RemoteFile` class. The latter has a `url` property but not a `status` property, whereas the former has `status` but not `url`. Clearly this code isn't right!

Flow doesn't complain, though. It thinks `this` has type `any` in the event listener.

Where possible, the DOM declarations should indicate the type of `this` in callbacks to prevent this sort of mistake.

For example, Google's Closure Compiler provides an [`@this`][1] annotation. You can see an example of this in action in the declaration for [`Array.prototype.find`][2].

[1]: https://developers.google.com/closure/compiler/docs/js-for-compiler#tag-this
[2]: https://github.com/google/closure-compiler/blob/816bcb75ca2d8c8e32dbb76f2164ad85cee5c1fb/externs/es6.js#L1096-L1104

**Comment by glortho** [comment]
In case you land here before this +1 functionality is added, and you can afford to do it, just do something like this:
```javascript
type Ctx = { foo: string, bar: number }
myMethod() {
  const myCtx: Ctx = this;
  // do stuff with myCtx instead of this
}
```

**Comment by STRML** [comment]
Just for reference, TS 2.0 [landed with this feature](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#specifying-the-type-of-this-for-functions) last month. It special-cases the first parameter if its name is `this`.

```javascript
function f(this: void) {
    // make sure `this` is unusable in this standalone function
}
interface UIElement {
    addClickListener(onclick: (this: void, e: Event) => void): void;
}
```

**Comment by nmn** [comment]
This already works in Flow:

```typescript
function head (count) {
  (this: Array<number>);
  return this.slice(0, count);
}

head.call([1, 2, 3]);
head.call(['123']); // error
```

**Comment by HerringtonDarkholme** [comment]
@nmn Yes we do need. 

It's for library declaration and use site completion. This is quite common for context injection.

```ts
function addListener<T: HTMLElement>(elem: T, func: (this: T) => void) {
  // implementation
}
addListener(htmlInputElement , function() {
  this.value // should have this completion
})
```

Another usage is in library like Vue/backbone.

```ts
Vue.extend({
  methods: {
    log() {
      this.instanceProperty // this is injected as Vue instance
    }
  }
})
```

Without this annotation, it is impossible for library author to provide a good type declaration for automatic completion.

**Comment by kazupon** [comment]
I hope this feature is provided 🙏

**Comment by scottbedard** [comment]
+1 for this feature

**Comment by iclanzan** [comment]
Has there been no progress on this yet?

**Comment by dmnd** [comment]
[@calebmer was looking for people to help implement it a while ago](https://twitter.com/calebmer/status/905051756157050880). I'm not sure if anyone started on it. I'd love to do it if I had more time.

**Comment by KindWizzard** [comment]
Any updates on this? :)

**Comment by akoppela** [comment]
@danvk It seems that you're using flow type version of `Q.Promise`, is it? If so do you mind to share a link to flow typed `Q.Pomise` library?

**Comment by danvk** [comment]
@akoppela https://github.com/hammerlab/pileup.js/blob/0182b225205e52e1489727004d69e84442eafb88/lib/q.js but take note that these are ~3 years old and there may be better ones available elsewhere.

**Comment by goodmind** [comment]
Hey! I tried to implement it at #7807

**Comment by Bannerets** [comment]
Note that `this` constraints in classes can be emulated via `$Call`

```js
declare class O<T> {
  m(): $Call<(string => string), T>;
  m(): $Call<(number => number), T>;
}
  
declare var o1: O<number>
  
const a1: number = o1.m() // ok
const b1: string = o1.m() // error

declare var o2: O<string>
  
const a2: number = o2.m() // error
const b2: string = o2.m() // ok

declare var o3: O<{}>
  
const a3 = o3.m() // error
const b3 = o3.m() // error
```

^ [flow try](https://flow.org/try/#0FAEwpgxgNghgTmABNGBnViDyAeAKgPkQG9hFEBbACgEoAuRAEgGEYoptLUAXOASwDsA5ogC8hbnyHUANIgIBuUhRr1mrdpX4BXcgCMwcUYW16DMufkUBfJaEiwEiAG7xEAewCM9HCf1x8thBu-NyIMF6Ivgai7h4AdFTUiAD0ye4A1sBBIVyIuhESAsIisQk0KWkGcG5wwHYoji6GbgBM3tiFQgFkWcGhMG2ROn4xrWVJqYhVNb05eYOdxe4t4xUZdeANSE3uAMztRFbdiLP9u6O7q5PTtdmhuuclbpeJazdAA)

```js
class M<+State> {
  +s: State
  constructor(s: State) { this.s = s }
  to2: $Call<(1 => () => M<2>) & () => {...}, State> = () => {
    return new M(2) // ok
  }
  to3: $Call<(1 => () => M<3>) & () => {...}, State> = () => {
    return new M(2) // error
  }
}

const m1 = new M(1)
const m2 = m1.to2() // ok
const m3 = m2.to2() // error
```

^ [flow try](https://flow.org/try/#0FAYwNghgzlAECyAeA1AZQC4XQUwHywG9hZZkoAuWDLbY2EAewDsp0AnAVxHQbYAoKVTDgCUhWOgAWASygA6OAF5YcAL50eAJkoASAMIQwYRHwCMsRfj5jLCRJtxiAZLGsX8BOV9UAaITXxlN1siEhI2bHQONiZYJmwAdwQ+TTEAejTYBgBrOnUSHgBmXQMjE3NbYPwkQsdYFyrCLzlffxxA1xsPOnDI6Nj4pPgU9MzsNjZePOB1UGZWWABbCrjE5NMROZZ0Jc0LJdM5LTcMrNzGbaXC-cXNI4ZNE7GJqaA)

```js
declare class O<+T> {
  constructor(T): void;
  m(): $Call<<U>(Array<U>) => U, T>
}
  
const oarr = new O([1,2,3])
const n: number = oarr.m() // ok
const s: string = oarr.m() // error
```

^ [flow try](https://flow.org/try/#0FAEwpgxgNghgTmABNGBnViDyAeA1AFQD5EBvYRZAewDtUAXOAVwjsrgAp8BKALkQDdKASxABucogC27XogAkAYRhQo2bAFVC7AIJw4MAJ4bCXRAF5i6gDSIiwAL4TgEGvUSV4cc4mpgA7ljsANoAjFYATFYAzAC6XM6udD581IySAEZgXmbungB00qYA9EXuANYJtEmofPRwQtQA5t4eegUyiCWIWXBswEA)

```js
declare class Ap<+T> {
  ap<A, B>(Ap<A>): $Call<(A => B) => Ap<B>, T>
}

const a = new Ap<number => string>
const n: Ap<number> = a.ap(new Ap<3>) // error
const s: Ap<string> = a.ap(new Ap<3>) // ok
```

^ [flow try](https://flow.org/try/#0FAEwpgxgNghgTmABNGBnViCCAHAPAagBUA+RAb2EURj0wBpEAhYgCh102IEoAuRAEgDCMKFFxtEAXlKMuU0u2YMSwAL7BgEAPYA7VABdqUxDrAB3LHh0BXALYAjMHHmIDcAJY6A5sU26DJnzsNg5OpJLUAHQ0LKYW7ADM3IgA9CmITnBacH56hqhBeG6ePsYw0dix5pa4SXJpiFoA1sBAA)

**Comment by iclanzan** [comment]
Wow, this is gold! Thank you @Bannerets!

**Comment by dmnd** [comment]
Fixing `this` was mentioned in the [latest roadmap](https://medium.com/flow-type/what-were-building-in-2020-bcb92f620c75).

**Comment by STRML** [comment]
The `this` annotation [was released in 0.139](https://github.com/facebook/flow/blob/master/Changelog.md#01390).

To use, add this to flowconfig:

```
[options]
experimental.this_annot`= true
```

To use:

```js
function foo(this: {toString: (number) => string}, a: number): string {
  return this.toString(a);
}
```

It is also automatically on at flow.org/try.

**Comment by gkz** [comment]
https://medium.com/flow-type/sound-typing-for-this-in-flow-d62db2af969e
